File: Lab8-Queue&Async/Lab8-Queue&Async/RaceVC/RaceVC.m
//
//  RaceVC.m
//  Lab8-Queue&Async
//
//  Created by 張力元 on 2019/3/15.
//  Copyright © 2019 張力元. All rights reserved.
//

#import "RaceVC.h"

@interface RaceVC ()
@property (strong, nonatomic) IBOutlet UIProgressView *rabbitProgress;
@property (strong, nonatomic) IBOutlet UIProgressView *tortoiseProgress;
@property (assign, nonatomic) double rabbitDis;
@property (assign, nonatomic) double tortoiseDis;
@property (assign, nonatomic) BOOL finishFlag;
@end

@implementation RaceVC
- (void)viewDidLoad {
    [super viewDidLoad];
    
}

- (IBAction)startBtnClick:(id)sender {
    //初始化烏龜、兔子走的距離與完成旗標
    _rabbitDis = 0;
    _tortoiseDis = 0;
    _finishFlag = false;
    self.rabbitProgress.progress = 0;
    self.tortoiseProgress.progress = 0;
    //dispatch_get_global_queue(佇列的優先權,旗標)
    dispatch_queue_t secQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0);
    dispatch_async(secQueue, ^{
        //兔子跑
        while(self.rabbitDis<100 &&
              self.tortoiseDis<100){
            [self rabbitRun];
            [self showProgress];
        }
        [self showWiner];
    });
    
    dispatch_async(secQueue, ^{
        //烏龜跑
        while(self.rabbitDis<100 &&
              self.tortoiseDis<100){
            [self tortoiseRun];
            [self showProgress];
        }
        [self showWiner];
    });
}

- (void)rabbitRun{
    //兔子走隨機(1~10) 休隨機(1~10)
    int move = arc4random()%10+1;
    _rabbitDis = _rabbitDis + move;
    int sleepTime = arc4random()%10+1;
    //如果烏龜跑超過八成則變成只休息1
    if (_tortoiseDis>80)
        sleepTime = 1;
    NSLog(@"兔子走%d休息%d",move,sleepTime);
    //設定休息時間 單位為微秒
    usleep(sleepTime*50000);
}

- (void)tortoiseRun{
    //烏龜穩定走二休一
    _tortoiseDis = _tortoiseDis + 2;
    usleep(50000);
}

- (void)showProgress{
    dispatch_queue_t mainQueue = dispatch_get_main_queue();
    //在主執行緒顯示畫面更新
    dispatch_async(mainQueue, ^{
        self.rabbitProgress.progress = self.rabbitDis/100;
        self.tortoiseProgress.progress = self.tortoiseDis/100;
    });
}

- (void)showWiner{
    if (!_finishFlag) {
        _finishFlag = true;
        NSString *winMsg;
        if (_rabbitDis > _tortoiseDis)
            winMsg = @"兔子勝利";
        else if (_tortoiseDis > _rabbitDis)
            winMsg = @"烏龜勝利";
        else
            winMsg = @"平手";
        dispatch_queue_t mainQueue = dispatch_get_main_queue();
        //在主執行緒顯示畫面更新
        dispatch_async(mainQueue, ^{
            UIAlertController *msgView = [UIAlertController alertControllerWithTitle:[NSString stringWithFormat:@"%@",winMsg] message:nil preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction *determineAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:nil];
            [msgView addAction:determineAction];
            [self presentViewController:msgView animated:true completion:nil];
        });
    }
}

@end

File: Swift/lab8/lab8-exercise/lab8-exercise/raceVC.swift
//
//  raceVC.swift
//  lab8-exercise
//
//  Created by taizhou on 2019/8/1.
//  Copyright © 2019 taizhou. All rights reserved.
//

import UIKit

class raceVC: UIViewController {
    @IBOutlet var rabbitProgress: UIProgressView!
    @IBOutlet var turtleProgress: UIProgressView!
    var rabbitDis = 0
    var turtleDis = 0
    var finishFlag = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    @IBAction func startBtnClick(_ sender: Any) {
        rabbitDis = 0
        turtleDis = 0
        finishFlag = false
        rabbitProgress.progress = 0
        turtleProgress.progress = 0
        let secQueue = DispatchQueue.global()
        secQueue.async {
            //兔子跑的任務，判斷是否有人抵達終點（100），若無則會繼續跑
            while(self.rabbitDis<100 && self.turtleDis<100){
                //呼叫「rabbitRun()」函式設置兔子每次的移動距離與休息時間
                self.rabbitRun()
                //呼叫「showProgress」函式顯示進度條的狀態
                self.showProgress()
            }
            //呼叫「showWinner」函式顯示勝負
            self.showWinner()
        }
        secQueue.async {
            //烏龜跑的任務，判斷是否有人抵達終點（100），若無則會繼續跑
            while(self.rabbitDis<100 && self.turtleDis<100){
                //呼叫「tortoiseRun()」函式設置烏龜每次的移動距離與休息時間
                self.turtletRun()
                //呼叫「showProgress」函式顯示進度條的狀態
                self.showProgress()
            }
            //呼叫「showWinner」函式顯示勝負
            self.showWinner()
        }
    }
    
    func rabbitRun(){
        //兔子走隨機(1~10) 休隨機(1~10)
        let move = Int.random(in: 1...10)
        rabbitDis+=move
        var sleepTime = Int.random(in: 1...10)
        //如果烏龜跑超過八成則變成只休息1倍時間
        if(turtleDis>80){
            sleepTime = 1
        }
        print("兔子走",move,"休息",sleepTime)
        usleep(useconds_t(sleepTime*50000)) //設定休息時間 單位為微秒
    }
    
    func turtletRun(){
        turtleDis+=2
        usleep(50000) //固定休息0.05秒
    }
    
    func showProgress(){
        let mainQueue = DispatchQueue.main
        mainQueue.async {
            self.rabbitProgress.progress = Float(self.rabbitDis)/100
            self.turtleProgress.progress = Float(self.turtleDis)/100
        }
    }
    
    func showWinner(){
        if(!finishFlag){
            finishFlag = true
            var winMsg = ""
            if(rabbitDis > turtleDis){
                winMsg = "兔子勝利"
            }
            else if(rabbitDis < turtleDis){
                winMsg = "烏龜勝利"
            }
            else{
                winMsg = "平手"
            }
            let mainQueue = DispatchQueue.main
            mainQueue.async {
                let msgView = UIAlertController(title: winMsg, message: "", preferredStyle: .alert)
                let determineAction = UIAlertAction(title: "確定", style: .default, handler: nil)
                msgView.addAction(determineAction)
                self.present(msgView, animated: true, completion: nil)
            }
        }
    }
}



